docs/source/tutorials/qmd/07-groups.qmd

```diff
@@ -0,0 +1,330 @@
+        self, name: str, a: float, b: float, start_value: float = 1000.0
+    ) -> None:
+        a_var = lsl.Var(a, name=f"{name}_a")
+        b_var = lsl.Var(b, name=f"{name}_b")
+
+        prior = lsl.Dist(tfd.InverseGamma, concentration=a_var, scale=b_var)
+        tau2 = lsl.Param(start_value, distribution=prior, name=name)
+        super().__init__(name=name, a=a_var, b=b_var, tau2=tau2)
+```
+
+
+### Spline coefficient
+
+Next, we create the group for our spline coefficient in a very similar way.
+
+
+```{python}
+class SplineCoef(lsl.Group):
+    def __init__(self, name: str, penalty: Array, tau2: lsl.Param) -> None:
+        penalty_var = lsl.Var(penalty, name=f"{name}_penalty")
+
+        # we save the rank of the penalty matrix here to use it later in our
+        # gibbs kernel
+        rank = lsl.Data(np.linalg.matrix_rank(penalty), _name=f"{name}_rank")
+
+        prior = lsl.Dist(
+            MultivariateNormalDegenerate.from_penalty,
+            loc=0.0,
+            var=tau2,
+            pen=penalty_var,
+        )
+        start_value = np.zeros(np.shape(penalty)[-1], np.float32)
+
+        coef = lsl.Param(start_value, distribution=prior, name=name)
+
+        super().__init__(name, coef=coef, penalty=penalty_var, tau2=tau2, rank=rank)
+
+
+```
+
+
+## Use groups
+
+We can use these groups to quickly create model building blocks and access their
+elements:
+
+```{python}
+tau2_group = VarianceIG(name="tau2", a=0.01, b=0.01)
+
+print(tau2_group["tau2"])
+```
+
+We can directly use the `tau2_group` to create an instance of
+`SplineCoef`:
+
+```{python}
+second_diff = np.diff(np.eye(10), n=2, axis=0)
+penalty = second_diff.T @ second_diff
+coef_group = SplineCoef(name="coef", penalty=penalty, tau2=tau2_group["tau2"])
+```
+
+Each group offers access to all its members through `mappingproxy` instances, which are
+basically immutable dictionaries:
+
+```{python}
+tau2_group.vars
+```
+
+```{python}
+coef_group.vars
+```
+
+## A complete P-Spline group
+
+In this last step, we create overarching group that incorporates the observed
+basis matrix of our spline and uses the parameter groups to evaluate the smooth
+function. Note the following:
+
+1. Here, we create a `SplineCoef` instance *inside* the new group's `__init__`, because
+   this step is always the same with P-Splines, once we know the penalty and the variance
+   parameter $\tau^2$.
+2. We do not automatically create an instance of our `IGVariance` group for the variance
+   parameter, but allow users to pass any fitting group instance to the class. This makes
+   our `PSpline` class very flexible: we could easily replace the `IGVariance` group
+   with another group that represents a different prior for $\tau^2$. For this `PSpline`
+   group, it is only important that there is a key `"tau2"` in the `tau2_group`.
+3. In our `super().__init__` call at the end, we make sure to add not only the
+   Vars `basis_matrix` and `smooth` that we directly created, but also all
```

Comment:
i like the term design matrix more than basis matrix but i can live with it.

i don't like the term smooth. can we use something different? for example f_of_x or effect_x

Reply:
I can see that, because the term "smooth" does not really speak to me, either. I chose it, because it fits with the terminology used throughout the `distreg` module. For the time being, I would tend to stick with `smooth`, but add an explanatory sentence.